### PR TITLE
LPD-39807 + LPD-39817

### PR DESCRIPTION
**Pull request by rijema**
Following pull with the changes requested here https://github.com/liferay-bpm/liferay-portal/pull/3833

LPD-39807 + LPD-39817

Related issues: 

- https://liferay.atlassian.net/browse/LPD-39807
- https://liferay.atlassian.net/browse/LPD-39817

### CODE REVIEW

**Comment by rijema** [comment]
ci:test:sf

**Comment by rijema** [comment]
Yes, it is everything working as it should :> 

image

image

image

image

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java [hunk 1]
```diff
@@ -416,6 +471,32 @@ public void testGetFullHierarchyDDMFormFieldsMap() throws Exception {
...
+
+		Assert.assertEquals(
+			structure,
+			_ddmStructureLocalService.getStructureByExternalReferenceCode(
+				structure.getExternalReferenceCode(), structure.getGroupId(),
+				structure.getClassNameId()));
+
+		_ddmStructureLocalService.deleteDDMStructure(
+			structure.getStructureId());
+	}
```
Comment (marcelabc) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java:498]:
```
@Test
	public void testGetStructureByExternalReferenceCode() throws Exception {
		String externalReferenceCode = RandomTestUtil.randomString();

		AssertUtils.assertFailure(
			NoSuchStructureException.class,
			StringBundler.concat(
				"No DDMStructure exists with the key {externalReferenceCode=",
				externalReferenceCode, ", groupId=", structure.getGroupId(),
				", classNameId=", structure.getClassNameId(), "}"),
			() -> _ddmStructureLocalService.getStructureByExternalReferenceCode(
				RandomTestUtil.randomString(), group.getGroupId(), _classNameId));

	      DDMStructure structure = _addStructure(
			      RandomTestUtil.randomString(), RandomTestUtil.randomString());

		Assert.assertEquals(
			structure,
			_ddmStructureLocalService.getStructureByExternalReferenceCode(
				structure.getExternalReferenceCode(), structure.getGroupId(),
				structure.getClassNameId()));

		_ddmStructureLocalService.deleteDDMStructure(
			structure.getStructureId());
	}
```

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java [hunk 2]
```diff
@@ -2289,6 +2343,28 @@ private void _validate(
...
+			externalReferenceCode, groupId, classNameId);
+
+		if ((structure != null) &&
+			(structure.getStructureId() != structureId)) {
+
+			throw new DuplicateDDMStructureExternalReferenceCodeException(
+				StringBundler.concat(
+					"Duplicate structure external reference code \"",
+					externalReferenceCode, "\" for class name ID \"",
+					classNameId, "\" in group \"", groupId, "\""));
```
Comment (marcelabc) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java:2364]: "Duplicate **DDM** structure external reference code \""

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/dto/v2_0/util/DataDefinitionUtil.java [hunk 3]
```diff
@@ -26,6 +27,18 @@
...
+		DataDefinitionContentType dataDefinitionContentType =
+			DataDefinitionContentTypeRegistryUtil.getDataDefinitionContentType(
+				structure.getClassNameId());
+
+		if (dataDefinitionContentType == null) {
+			return null;
+		}
+
+		return dataDefinitionContentType.getContentType();
+	}
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/dto/v2_0/util/DataDefinitionUtil.java:40]: Why did you create another method? Are you using in another place?
If not, please revert as it was.

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/resource/v2_0/DataDefinitionResourceImpl.java [hunk 4]
```diff
@@ -509,6 +562,27 @@ public DataDefinition putDataDefinition(
...
 
+	@Override
+	public DataDefinition
+			putSiteDataDefinitionByContentTypeByExternalReferenceCode(
+				Long siteId, String contentType, String externalReferenceCode,
+				DataDefinition dataDefinition)
+		throws Exception {
+
+		DDMStructure ddmStructure =
+			_ddmStructureLocalService.getStructureByExternalReferenceCode(
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/resource/v2_0/DataDefinitionResourceImpl.java:573]: can you change it for the `fetch` method?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/model/DDMFormField.java [hunk 5]
```diff
@@ -53,6 +50,14 @@ public DDMFormField(DDMFormField ddmFormField) {
 			addDDMFormFieldRule(new DDMFormFieldRule(ddmFormFieldRule));
 		}
 
+		DDMFormFieldOptions ddmFormFieldOptions =
+			ddmFormField.getDDMFormFieldOptions();
+
+		if (ddmFormFieldOptions != null) {
+			setDDMFormFieldOptions(
+				new DDMFormFieldOptions(ddmFormFieldOptions));
+		}
```
Comment (marcelabc) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/model/DDMFormField.java:59]: why did you move this down? can we maintain this code in the original line?

**Review by marcelabc — CHANGES_REQUESTED** [review]
Hi 
I finished my review. Let me know if you have questions :smile:

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/exception/NoSuchDataDefinitionException.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.data.engine.rest.internal.exception;
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/exception/NoSuchDataDefinitionException.java:6]: this exception should be in the -api module. Please, search for similar uses.
you should create this class in `data-engine-api` `com.liferay.data.engine.exception`

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/model/DDMFormField.java (hunk 5 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/model/DDMFormField.java:59]: Yes, we can! I tried to put closer to the others null verification and sets we have on this method.

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/exception/NoSuchDataDefinitionException.java (hunk 6 above)
Comment (rijema) [reply modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/exception/NoSuchDataDefinitionException.java:6]: Ok doing it.

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java [hunk 7]
```diff
@@ -125,6 +125,15 @@ public void testDeleteDataDefinition() throws Exception {
 				parentDataDefinition.getId()));
 	}
 
+	@Override
+	@Test
+	public DataDefinition
+			testDeleteSiteDataDefinitionByContentTypeByExternalReferenceCode_addDataDefinition()
+		throws Exception {
+
+		return _addDataDefinition("data-definition-1.json");
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:134]: can you change this return for this:
```
return dataDefinitionResource.postSiteDataDefinitionByContentType(
			testGroup.getGroupId(), _CONTENT_TYPE, randomDataDefinition());
```

Comment (marcelabc) [reply modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:134]: See this example: `testPutDataDefinitionPermissionsPage_addDataDefinition()`

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java [hunk 8]
```diff
@@ -150,6 +159,15 @@ public void testGetDataDefinitionPermissionsPage() throws Exception {
 		Assert.assertNotNull(page);
 	}
 
+	@Override
+	@Test
+	public DataDefinition
+			testGetSiteDataDefinitionByContentTypeByExternalReferenceCode_addDataDefinition()
+		throws Exception {
+
+		return _addDataDefinition("data-definition-1.json");
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:168]: can you apply the last comment here please?

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java [hunk 9]
```diff
@@ -705,6 +748,24 @@ public void testPutDataDefinition() throws Exception {
 		assertValid(getDataDefinition);
 	}
 
+	@Override
+	@Test
+	public DataDefinition
+			testPutSiteDataDefinitionByContentTypeByExternalReferenceCode_addDataDefinition()
+		throws Exception {
+
+		return _addDataDefinition("data-definition-1.json");
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:757]: can you apply the last comment here please?

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java [hunk 10]
```diff
@@ -886,6 +947,12 @@ protected DataDefinition testPutDataDefinition_addDataDefinition()
 			testGroup.getGroupId(), _CONTENT_TYPE, randomDataDefinition());
 	}
 
+	private DataDefinition _addDataDefinition(String file) throws Exception {
+		return dataDefinitionResource.postSiteDataDefinitionByContentType(
+			testGroup.getGroupId(), _CONTENT_TYPE,
+			DataDefinition.toDTO(DataDefinitionTestUtil.read(file)));
+	}
+
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:955]: I believe we can remove this method.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java (hunk 2 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java:2364]: As we're inside the service layer where there is only one kind of structure, I thought we would consider maintain this pattern without the prefix DDM. Doing the change rn

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/dto/v2_0/util/DataDefinitionUtil.java (hunk 3 above)
Comment (rijema) [reply modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/dto/v2_0/util/DataDefinitionUtil.java:40]: Yes 

https://github.com/liferay-bpm/liferay-portal/blob/a20b3326e0c09650b202f3e3b4edd657c0352541/modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/dto/v2_0/util/DataDefinitionUtil.java#L60

https://github.com/liferay-bpm/liferay portal/blob/a20b3326e0c09650b202f3e3b4edd657c0352541/modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/resource/v2_0/DataDefinitionResourceImpl.java#L1360

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/resource/v2_0/DataDefinitionResourceImpl.java (hunk 4 above)
Comment (rijema) [reply modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/resource/v2_0/DataDefinitionResourceImpl.java:573]: Yes, we can. Doing as soon as possible

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java (hunk 7 above)
Comment (rijema) [reply modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:134]: Sure, np

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java (hunk 9 above)
Comment (rijema) [reply modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:757]: Sure

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java [hunk 11]
```diff
@@ -705,6 +748,24 @@ public void testPutDataDefinition() throws Exception {
...
+		return _addDataDefinition("data-definition-1.json");
+	}
+
+	@Override
+	@Test
+	public DataDefinition
+			testPutSiteDataDefinitionByContentTypeByExternalReferenceCode_createDataDefinition()
+		throws Exception {
+
+		return testGetDataDefinition_addDataDefinition();
```
Comment (marcelabc) [inline modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:766]: can you change the return to:
```
return dataDefinitionResource.postSiteDataDefinitionByContentType(
			testGroup.getGroupId(), _CONTENT_TYPE, randomDataDefinition());
```

File: modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java (hunk 10 above)
Comment (rijema) [reply modules/apps/data-engine/data-engine-rest-test/src/testIntegration/java/com/liferay/data/engine/rest/resource/v2_0/test/DataDefinitionResourceTest.java:955]: Ok

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/model/DDMFormField.java (hunk 5 above)
Comment (marcelabc) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/model/DDMFormField.java:59]: lets maintain.

Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/model/DDMFormField.java:59]: Yes ma'am

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java (hunk 2 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java:2364]: Done

**Comment by rijema** [comment]
ci:test:sf

**Comment by rijema** [comment]
ci:test:relevant

**Comment by marcelabc** [comment]
resend: https://github.com/liferay-bpm/liferay-portal/pull/3843
